src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -256,6 +376,132 @@ class GeometricInt64Generator : public Generator<int64_t> {
     const double _p;
 };
 
+// This generator allows choosing any valid generator, incuding documents. As such it cannot be used
+// by JoinGenerator today. See ChooseStringGenerator.
+class ChooseGenerator : public Appendable {
+public:
+    // constructore defined at bottom of the file to use other symbol
+    ChooseGenerator(const Node& node, DefaultRandom& rng);
+    Appendable& choose() {
+        // Pick a random number between 0 and sum(weights)
+        // Pick value based on that.
+        auto distribution = boost::random::discrete_distribution(_weights);
+        return (*_choices[distribution(_rng)]);
+    }
+
+    void append(const std::string& key, bsoncxx::builder::basic::document& builder) override {
+        choose().append(key, builder);
+    }
+    void append(bsoncxx::builder::basic::array& builder) override {
+        choose().append(builder);
+    }
+
+protected:
+    DefaultRandom& _rng;
+    std::vector<UniqueAppendable> _choices;
```

Comment:
My C++ skills aren't up to this -- but maybe...
Is there a way to definite a template function the exist only if it can exist? I'd like to add a stringEvaluate function to Generator<T>, but only if T can be passed as an argument to to_string http://www.cplusplus.com/reference/string/to_string/ or some similar form. Ideally we could handle the  bson view version also.

Reply:
That is what I was looking for, but also scary :) One thing this would enable (elsewhere) would be joining not just string types, but also numbers easily. I think I figured out how to do this without the enable_if stuff, but also resisting the urge to do it as it's not worth the effort right now. I think I could add a generateString method to Generator<T>, and have it call to_string(generate()), and then I just need to make sure to define a to_string for every T used with Generator. 

It's not really worth it, especially after I realized I can generate random strings that look like numbers easily enough through the alphabet option.